File: MyLeetCode_Go/study/LeetCode/Problem_260_SingleNumber.go
package main

func singleNumber(nums []int) []int {
	if nums == nil || len(nums) == 0 {
		return []int{}
	}
	var xor = 0
	for _, num := range nums {
		xor ^= num
	}
	// 提取最右侧的 1, 将数字分成两类
	mostRightOne := xor & (^xor + 1)
	var oneNum = 0
	for _, num := range nums {
		// 找出带 1 的哪一类数字
		if num&mostRightOne != 0 {
			oneNum ^= num
		}
	}
	return []int{oneNum, xor ^ oneNum}
}

File: MyLeetCode_Java/src/main/java/LeetCode/Problem_260_SingleNumber.java
package LeetCode;

public class Problem_260_SingleNumber {

    // 给定一个整数数组 nums，其中恰好有两个元素只出现一次，
    // 其余所有元素均出现两次。 找出只出现一次的那两个元素。
    // 你可以按 任意顺序 返回答案。
    // 异或，英文为exclusive OR，缩写成xor异或（xor）
    public int[] singleNumber(int[] nums) {

        if (nums == null || nums.length == 0) {
            return new int[]{};
        }

        int xor = 0;
        for (int num: nums) {
            xor ^= num;
        }
        // 获取最右侧的 1
        int mostRightOne = xor & (-xor);
        int oneXor = 0;
        for (int num: nums) {
            if ((num & mostRightOne) != 0) {
                oneXor ^= num;
            }
        }
        return new int[]{oneXor, xor^oneXor};
    }
}
